text/0066-separate-context-propagation.md

```diff
@@ -0,0 +1,499 @@
+  extract = Propagation::GetHTTPExtractor()
+  extract(request.Headers)
+  
+  Tracer::StartSpan([span options])
+  
+  version = Baggage::GetBaggage("client-version")
+  
+  switch( version ){
+    case "v1.0":
+      data = FetchDataFromServiceB()
+    case "v2.0":
+      data = FetchDataFromServiceC()
+  }
+
+  request.Response(data)
+  
+  Tracer::EndSpan()
+}
+
+func FetchDataFromServiceB() -> (data) {
+  request = newRequest([request options])
+  
+  inject = Propagation::GetHTTPInjector()
+  inject(request.Headers)
+  
+  data = request.Do()
+
+  return data
+}
+```
+
+Digging into the details of the tracing system, what might some of the details 
+look like? Here is a crude example of extracting and injecting B3 headers, 
+using an explicit context.
+
+```php
+  func B3Extractor(context, headers) -> (context) {
+    context = Context::SetValue( context, 
+                                 "trace.parentTraceID", 
+                                 headers["X-B3-TraceId"])
+    context = Context::SetValue( context,
+                                "trace.parentSpanID", 
+                                 headers["X-B3-SpanId"])
+    return context
+  }
+
+  func B3Injector(context, headers) -> (context) {
+    headers["X-B3-TraceId"] = Context::GetValue( context, "trace.parentTraceID")
+    headers["X-B3-SpanId"] = Context::GetValue( context, "trace.parentSpanID")
+
+    return context
+  }
+```
+
+Now, have a look at a crude example of how StartSpan might then make use of the 
+context. Note that this code must know the internal details about the context 
+keys in which the propagators above store their data. For this pseudocode, let's 
+assume again that the context is passed implicitly in a thread local.
+
+```php
+  func StartSpan(options) {
+    spanData = newSpanData()
+    
+    spanData.parentTraceID = Context::GetValue( "trace.parentTraceID")
+    spanData.parentSpanID = Context::GetValue( "trace.parentSpanID")
+    
+    spanData.traceID = newTraceID()
+    spanData.spanID = newSpanID()
+    
+    Context::SetValue( "trace.parentTraceID", spanData.traceID)
+    Context::SetValue( "trace.parentSpanID", spanData.spanID)
+    
+    // store the spanData object as well, for in-process propagation. Note that 
+    this key will not be propagated downstream.
+    Context::SetValue( "trace.currentSpanData", spanData)
+
+    return
+  }
+```
+
+# Internal details
+
+![drawing](img/context_propagation_details.png)
+
+## Context details
+OpenTelemetry currently intends to implement three context types of context 
+propagation.
+
+**Span Context -** The serializable portion of a span, which is injected and 
+extracted. The readable attributes are defined to match those found in the 
+[W3C Trace Context specification](https://www.w3.org/TR/trace-context/). 
+
+**Correlation Context -** Correlation Context contains a map of labels and 
+values, to be shared between metrics and traces. This allows observability data 
```

Comment:
I presume logs are also expected in the future? might just be better to say `to be shared across different types of telemetry (such as metrics and traces)`.

Reply:
sounds good. Side comment that clarifying correlation context might itself be an OTEP, given the number of discussions around it :)